application-migrator-migrators/application-migrator-migrators-property/application-migrator-migrators-property-default/src/main/java/org/xwiki/contrib/migrator/migrators/internal/PropertyMigrationExecutor.java

```diff
@@ -0,0 +1,251 @@
+
+    @Inject
+    @Named("context")
+    private ComponentManager componentManager;
+
+    static class PropertyConversionCheckerProvider
+    {
+        private List<PropertyConversionChecker> checkList;
+
+        PropertyConversionCheckerProvider()
+        {
+            this.checkList = new ArrayList<PropertyConversionChecker>();
+        }
+
+        public List<PropertyConversionChecker> getList()
+        {
+            return this.checkList;
+        }
+    }
+
+    private class PropertyConversionChecker
+    {
+        private List<String> allSupportedProperties;
+
+        private String propertyType;
+
+        PropertyConversionChecker(String propertyType)
+        {
+            this.propertyType = propertyType;
+        }
+
+        public List<String> getAllSupportedProperties()
+        {
+            return this.allSupportedProperties;
+        }
+
+        public String getPropertyType()
+        {
+            return this.propertyType;
+        }
+    }
+
+    @Override
+    public MigrationStatus execute(PropertyMigrationDescriptor migrationDescriptor) throws MigrationException
+    {
+        this.migrationDescriptor = migrationDescriptor;
+        migrationParameters = (PropertyMigrationParameters) this.migrationDescriptor.getMigrationParameters();
+
+        xWikiContext = xWikiContextProvider.get();
+        xwiki = xWikiContext.getWiki();
+
+        populizePropertyConversionChecker();
+
+        // STEP 1 : we get two properties (old property and new property could be document meta data)
+        classReference = documentReferenceResolver.resolve(migrationParameters.getTargetClass());
+        if (!xwiki.exists(classReference, xWikiContext)) {
+            logger.error("The new class reference does not exists! Aborting ...");
+            throw new MigrationException("Failed to migrate the XClasses : the new class does not exist.");
+        }
+
+        BaseClass c = new BaseClass();
+
+        try {
+            c = xwiki.getDocument(classReference, xWikiContext).getXClass();
+        } catch (XWikiException e) {
+            e.printStackTrace();
+        }
+
+        // get two propertyClasses, then compare name
+        for (Object o : c.getProperties()) {
+            PropertyClass propertyClass = (PropertyClass) o;
+            if (propertyClass.getName().equals(migrationParameters.getOldProperty())) {
+                oldPropertyType = propertyClass;
+            }
+            if (propertyClass.getName().equals(migrationParameters.getNewProperty())) {
+                newPropertyType = propertyClass;
+            }
+        }
+        // STEP 2 : check compatibility (from anything to String (Compatible))
+        if (areTwoPropertiesCompatible(oldPropertyType, newPropertyType)) {
+            // STEP 3: start migration
+            // TODO : isRemovedOldProperty?
+            try {
+                migrateProperty(xwiki.getDocument(classReference, xWikiContext));
```

Comment:
We are currently migrating only the properties of the `classReference` document. Let's consider the following conditions : 
* `MyProject.MyClass` contains : 
** The definition of the class that we want to migrate
** Some objects of the `MyProject.MyClass` XClass
* `MyProject.MyDoc1` contains some objects of the `MyProject.MyClass` XClass
* `MyProject.MyDoc2` contains some objects of the `MyProject.MyClass` XClass

Using the current code, only the first object of `MyProjet.MyClass` will be migrated. We need to actually gather all the documents in the current wiki that contain objects of the `MyProject.MyClass` XClass.

For that, you can use HQL queries, see https://extensions.xwiki.org/xwiki/bin/view/Extension/Query%20Module. I think that we also have a similar implementation in the ClassMigrator.